projects/notional/index.js

```diff
@@ -1,61 +1,23 @@
+const { sumTokens2 } = require('../helper/unwrapLPs');
 const abi = require('./abi');
-const sdk = require('@defillama/sdk');
-const { sumTokensAndLPsSharedOwners } = require('../helper/unwrapLPs');
 
-const escrowContract = '0x9abd0b8868546105F6F48298eaDC1D9c82f7f683';
 const v2Contract = "0x1344A36A1B56144C3Bc62E7757377D288fDE0369"
 
-async function tvl (timestamp, block) {
-  const maxCurrencyId = (await sdk.api.abi.call({
-    block,
-    target: escrowContract,
-    abi: abi['maxCurrencyId']
-  })).output;
-
-  const addressCalls = []
-  for (let i = 0; i <= maxCurrencyId; i++) {
-    addressCalls.push({
-      target: escrowContract,
-      params: i
-    })
-  }
-
-  const supportedTokens = (await sdk.api.abi.multiCall({
-    calls: addressCalls,
-    target: escrowContract,
-    abi: abi['currencyIdToAddress'],
-    block,
-  })).output
-
-  const balanceCalls = supportedTokens.map((s) => {
-    return {
-      target: s.output,
-      params: escrowContract
-    }
-  })
-
-  const balances = (await sdk.api.abi.multiCall({
-    calls: balanceCalls,
-    abi: 'erc20:balanceOf',
-    block
-  })).output
-
-  const balanceMap = balances.reduce((obj, b) => {
-    obj[b.input.target] = b.output
-    return obj
-  }, {})
-
-  await sumTokensAndLPsSharedOwners(balanceMap, [
-    "0x39aa39c021dfbae8fac545936693ac917d5e7563",
-    "0x5d3a536E4D6DbD6114cc1Ead35777bAB948E3643",
-    "0x4ddc2d193948926d02f9b1fe9e1daa0718270ed5",
-    "0xccf4429db6322d5c611ee964527d42e5d685dd6a"
-  ].map(a=>[a, false]), [v2Contract], block)
-
-  return balanceMap
+async function tvl(timestamp, block, _, { api }) {
+  let tokens = await api.fetchList({ lengthAbi: abi.getMaxCurrencyId, itemAbi: abi.getCurrency, target: v2Contract, startFromOne: true, })
+  tokens = tokens.flat().map(i => i[0])
+  const tokenNames = await api.multiCall({  abi: 'string:name', calls: tokens, permitFailure: true, })
+  const nwTokens = tokens.filter((v, i) => tokenNames[i] && tokenNames[i].startsWith('Notional Wrapped'))
+  let nwBals = await api.multiCall({  abi: 'erc20:balanceOf', calls: nwTokens.map(i => ({ target: i, params: v2Contract}))})
+  const underlyingTokens = await api.multiCall({  abi: 'address:underlying', calls: nwTokens})
+  const exchangeRate = await api.multiCall({  abi: 'uint256:getExchangeRateView', calls: nwTokens})
+  const tDecimals = await api.multiCall({  abi: 'erc20:decimals', calls: nwTokens})
+  const uDecimals =( await api.multiCall({  abi: 'erc20:decimals', calls: nwTokens, permitFailure: true,})).map(i => i ?? 18)
+  nwBals = nwBals.map((bal, i) => bal * (exchangeRate[i]/1e18) * (10 ** uDecimals[i] / 10 ** tDecimals[i]))
```

Comment:
The output value you have looks right although I don't know if you actually need to convert decimal basis twice here since the decimal basis conversion is already wrapped up in the exchange rate: https://github.com/notional-finance/contracts-v2/blob/master/contracts/external/adapters/nwToken.sol#LL191C2-L193C6

I don't see a problem with how you're doing it here since the output looks right.
```
   function _convertToUnderlying(uint256 assetAmount) private view returns (uint256) {
        return assetAmount * finalExchangeRate / EXCHANGE_RATE_PRECISION;
    }
    ```

Reply:
ah, didnt realize that I nearly introduced a bug, it was supposed to be nwToken decimals and underlying token decimals, but called against nwTokens tice by mistake.